2-WebApp-graph-user/2-5-HybridFlow/Views/Shared/_Layout.cshtml

```diff
@@ -0,0 +1,226 @@
+    <link rel="stylesheet" href="~/css/site.css" />
+    <script src="https://alcdn.msauth.net/browser/2.22.0/js/msal-browser.min.js"></script>
+    <script>
+        const msalInstance = new msal.PublicClientApplication({
+            auth: {
+                @{
+                    var clientId = Configuration["AzureAd:ClientId"];
+                    var instance = Configuration["AzureAd:Instance"];
+                    var domain = Configuration["AzureAd:Domain"];
+                    var redirectUri = Configuration["SpaRedirectUri"];
+
+                    @Html.Raw($"clientId: '{clientId}',");
+                    @Html.Raw($"redirectUri: '{redirectUri}',");
+                    @Html.Raw($"authority: '{instance}{domain}',");
+                }
+            },
+            cache: {
+                cacheLocation: 'sessionStorage',
+                    storeAuthStateInCookie: false,
+                }
+            });
+
+        function getTokenFromCache(scopes) {
+            @if (User.Identity is not null)
+            {
+                @Html.Raw($"const username = '{User.Identity.Name}';");
+            }
+            else
+            {
+                @Html.Raw($"const username = '';");
+            }
+
+            const account = msalInstance.getAllAccounts()
+                .find(account => account.username === username);
+
+            return msalInstance.acquireTokenSilent({
+                account,
+                scopes
+            })
+                .catch(error => {
+                    if (error instanceof msal.InteractionRequiredAuthError) {
+                        return msalInstance.acquireTokenRedirect({
+                            account,
+                            scopes
+                        });
+                    }
+                });
+        };
+
+        function callMSGraph(path, token) {
+            @{
+                var graphEndpoint = Configuration["DownstreamApi:BaseUrl"];
+                if (!string.IsNullOrEmpty(graphEndpoint))
+                {
+                    @Html.Raw($"const graphEndpoint = '{graphEndpoint}';");
+                }
+                else
+                {
+                    @Html.Raw($"const graphEndpoint = '';");
+                }
+            }
+
+            const headers = new Headers();
+            const bearer = `Bearer ${token}`;
+            headers.append("Authorization", bearer);
+
+            const options = {
+                method: "GET",
+                headers
+            };
+
+            console.log('request made to Graph API at: ' + new Date().toString());
+
+            return fetch(`${graphEndpoint}${path}`, options)
+                .catch(error => console.log(error))
+        }
+
+        async function signOut() {
+            await msalInstance.logoutRedirect({
+                onRedirectNavigate: _ => false
+            });
+
+            window.location = 'https://localhost:7089/MicrosoftIdentity/Account/SignOut';
+        }
```

Comment:
@jasonnutter 

Do you have any feedback here?

Unfortunately I can't seem to find any realistic solution for clearing the local cache without terminating the logout redirect from `msal.js` and then linking to the server logout after it finishes. Not sure if this would be the cleanest path.

Reply:
I would inverse the pattern here: When teh users click the button to sign out, redirect them to a page (e.g. /MicrosoftIdentity/Account/SignOut) that when loaded server side, clears server side state (without redirecting to AAD), and then client side loads MSAL and invokes `loginRedirect` and lets it redirect to AAD (ie don't implement `onRedirectNavigate`).